docs/source/user-guide/arrow-introduction.md

```diff
@@ -0,0 +1,284 @@
+            Arc::new(StringArray::from(vec![Some("foo"), Some("bar"), None])) as _,
+        ],
+    )?;
+
+    // expose it as a table
+    let table = MemTable::try_new(schema, vec![vec![batch]])?;
+    ctx.register_table("people", Arc::new(table))?;
+
+    // query it
+    let df = ctx.sql("SELECT id, upper(name) AS name FROM people WHERE id >= 2").await?;
+    df.show().await?;
+    Ok(())
+}
+```
+
+## Common Pitfalls
+
+When working with Arrow and RecordBatches, watch out for these common issues:
+
+- **Schema consistency**: All batches in a stream must share the exact same [`Schema`]. For example, you can't have one batch where a column is [`Int32`] and the next where it's [`Int64`], even if the values would fit
+- **Immutability**: Arrays are immutable—to "modify" data, you must build new arrays or new RecordBatches. For instance, to change a value in an array, you'd create a new array with the updated value
+- **Buffer management**: Variable-length types (UTF-8, binary, lists) use offsets + values arrays internally. Avoid manual buffer slicing unless you understand Arrow's internal invariants—use Arrow's built-in compute functions instead
+- **Type mismatches**: Mixed input types across files may require explicit casts. For example, a string column `"123"` from a CSV file won't automatically join with an integer column `123` from a Parquet file—you'll need to cast one to match the other
+- **Batch size assumptions**: Don't assume a particular batch size; always iterate until the stream ends. One file might produce 8192-row batches while another produces 1024-row batches
+
+## When Arrow knowledge is needed (Extension Points)
+
+For many use cases, you don't need to know about Arrow. DataFusion handles the conversion from formats like CSV and Parquet for you. However, Arrow becomes important when you use DataFusion's **[extension points]** to add your own custom functionality.
+
+These APIs are where you can plug your own code into the engine, and they often operate directly on Arrow [`RecordBatch`] streams.
+
+- **[`TableProvider`] (Custom Data Sources)**: This is the most common extension point. You can teach DataFusion how to read from any source—a custom file format, a network API, a different database—by implementing the [`TableProvider`] trait. Your implementation will be responsible for creating [`RecordBatch`]es to stream data into the engine. See the [Custom Table Providers guide] for detailed examples.
+
+- **[User-Defined Functions (UDFs)]**: If you need to perform a custom transformation on your data that isn't built into DataFusion, you can write a UDF. Your function will receive data as Arrow arrays (inside a [`RecordBatch`]) and must produce an Arrow array as its output.
+
+- **[Custom Optimizer Rules and Operators]**: For advanced use cases, you can even add your own rules to the query optimizer or implement entirely new physical operators (like a special type of join). These also operate on the Arrow-based query plans.
```

Comment:
I think typically custom optimizer rules are more concerned about the Schemas than the arrays, but I think we can leave this as is too